### PR TITLE
CDAP-13723: Make error message for metadata user friendly

### PR DESCRIPTION
**Pull request by rohitsinha54**
## Summary
- In CDAP 5.0 we introduced MetadataEntity which internally uses a LinkedHashMap to store the key-value pair for resource. For example a dataset looks like this:
MetadataEntity{details={namespace="ns1", dataset="ds1"}, type="dataset"}  this shows up in UI to user if a metadata calls fails as the error message.
- This PR makes the representation which is show to user for MetadataEntity more friendly.
    1. For EntityId (i.e. CDAP Entities) we will show user helpful descriptive messages for example
    > Unable to set metadata for application <app-name> in namespace <ns-name>  : xxxxxx should 
       not exceed 50 characters.
 
    2. For custom entities we show the list of hierarchical key-values and not the implementation details such as the data structure representation etc.

## Misc
[Issue](https://issues.cask.co/browse/CDAP-13723)
Build: https://builds.cask.co/browse/CDAP-RUT1534-1

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java [hunk 1]
```diff
@@ -277,6 +277,110 @@ public static MetadataEntity ofDataset(String namespace, String datasetName) {
...
+  /**
+   * @return a {@link String} which describes the {@link MetadataEntity} for a user in plain english.
+   * If the {@link MetadataEntity} represents a known CDAP entity then the description is worded to show relations
+   * in the hierarchy.
+   */
+  public String getDescription() {
+    return getDescription(new StringBuilder(), getType());
+  }
+
+  private String getDescription(StringBuilder builder, String type) {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java:292]: It seems like it would be more readable code if you used `String.format` and String concatenation where necessary, instead of using StringBuilder like you are currently doing.
The performance won't be as good, but it doesn't seem like we need to make `getDescription` super performant.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java:292]: Agreed it will make it much more readable. Yes, we dont accept this call to be infrequent.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java [hunk 2]
```diff
@@ -277,6 +277,79 @@ public static MetadataEntity ofDataset(String namespace, String datasetName) {
...
+      case MetadataEntity.NAMESPACE:
+        builder.append(String.format("%s: %s", MetadataEntity.NAMESPACE, getValue(MetadataEntity.NAMESPACE)));
+        return builder.toString();
+      case MetadataEntity.APPLICATION:
+        String description = String.format("%s: %s", MetadataEntity.APPLICATION, getValue(MetadataEntity.APPLICATION));
+        // MetadataEntity can be created without version and in that case cdap treats it as default version
+        // internally but for no surprises to user no need to display version if user specifically created this
+        // without version. If the cdap system has created the MetadataEntity it will have version info set even for
+        // default version
+        if (containsKey(MetadataEntity.VERSION)) {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java:303]: how come we say "app ... deployed in ... ns" only if it contains version key?
Shouldn't we still say "deployed in" otherwise?

**Review by anwar6953 — APPROVED** [review]
